### ISSUE TITLE
Discussion: How to keep ItemsRepeater's all items

### ISSUE THREAD

**Issue by hupo376787** [microsoft/microsoft-ui-xaml#3861]
# Discussion: [Keep ItemsRepeater's all items] 
Suppose an `ItemsRepeater` is vertical, and it has 50+ items. If I scrolling down to the bottom, the top items maybe dispose of and regenerate when scrolling to the top.

I set the `VerticalCacheLength` to a number, like 100. But it seems the same things happen again.

So how to keep all items.

**Comment by huoyaoyuan** [comment]
This is a feature called UI virtualization. To avoid this, use a non-virtualizing `Layout`, or use `ItemsControl` with non-virtualizing `ItemsPanel`.

With that said, it's suggested to keep all required state in your view model. Control is much heavier than view model.

**Comment by hupo376787** [comment]
I used `StackPanel` to prevent virtualizing, In each item data template, I also used an `ItemsRepeater`, so it likes an matrix generally.

If I scroll the first row to the last item, and then scroll to the last row, then scroll back, the first row is scrolled to the 1st item.
I can use a picture to describle.


![image](https://user-images.githubusercontent.com/14960168/103163389-549ee500-4838-11eb-90a7-de1c74179a1a.png)

**Comment by hupo376787** [comment]
So I just want to keep the all rows view state.

**Comment by huoyaoyuan** [comment]
I'm not sure the behavior here. But you may create a view model for each row to hold its selection state.

**Comment by maxkatz6** [comment]
> I used StackPanel to prevent virtualizing, In each item data template

If you want to disable virtualization with StackPanel, you need to wrap whole ItemsRepeater with it, so repeater will be measured with infinity height. 
But I agree with @huoyaoyuan that you should keep all important for you state in the ViewModel.

**Comment by hupo376787** [comment]
You mean I need to keep each row's selected item/index, when the row start to bring into view, then scroll to the item/index.

Am I right?

**Comment by ranjeshj** [comment]
@hupo376787 Couple of options. If you really need to not virtualize, you can write a custom non-virtualizing layout for ItemsRepeater (should be simple for Stacking layout). However, if you have reasonably large number or items, performance will take a big hit (hence the need for UI Virtualization). Keeping track of the selection state in the view model as @huoyaoyuan suggested seems like a reasonable way to avoid it and get good perf.

**Comment by hupo376787** [comment]
And recently we found items that may disappear randomly using up/down key to focus.

My xaml [gist](https://gist.github.com/hupo376787/496aee061296136102bcb66037b84054).

https://user-images.githubusercontent.com/14960168/104161564-0b50b700-542e-11eb-990b-b576d0bb4ea6.mp4

**Comment by hupo376787** [comment]
This problem is caused by we set `ItemsRepeater`'s `Visibility` to collapse when data is loading, now we set `ItemsRepeater`'s `Visibility` always visible.

But there are still other bugs.
1. The `Windows.UI.Xaml.Controls. Image` goes totally black sometimes like below.
![image](https://user-images.githubusercontent.com/14960168/104543917-8c43c480-5661-11eb-83be-daf406c2ad2d.png)
2. Updating the vertical `ItemsRepeater ItemsSource` may cause horizontal `ItemsRepeater` disappear.
![image](https://user-images.githubusercontent.com/14960168/104544266-4affe480-5662-11eb-8ebd-02c366ad709d.png)
